### PR TITLE
WebGLRender & NodeMaterials program caching errors

### PR DESCRIPTION
**Pull request by aardgoose**
**  Program caching issue with Node materials and the webgl renderer **

With certain NodeMaterials eg MeshBasicNodeMaterial and MeshPhongNodeMaterial, program caching can fail with various breakages: example test case added in the PR: examples\webgl_nodes_test.html based on webgl_nodes_materials_physical_clearcoat.html.

The root cause is that program parameters from programCache.getParameters() are relied on to generate a unique key value, however the parameters and key are derived in turn from the material's shaders. 

In the NodeMaterials which derive from ShaderMaterial, all node materials have the default ShaderMaterial shaders at this point, and the real shaders are only built later via onBuild(). Depending on the material's other parameters it is possible to get duplicate cache keys for different NodeMaterials, and hence the wrong programs being used for some materials.

I have have removed the inherited shaders in NodeMaterial and moved the material.onBuild() call into getParameters() to ensure accurate custom<stage>ShaderID values are generated.

I am not sure if this is the best solution, but it fixes the problem.
 
@sunag

### CODE REVIEW

File: examples/jsm/nodes/materials/NodeMaterial.js [hunk 1]
```diff
@@ -44,6 +44,9 @@ class NodeMaterial extends ShaderMaterial {
 
 		this.positionNode = null;
 
+		this.vertexShader = null;
```
Comment (LeviPesin) [inline examples/jsm/nodes/materials/NodeMaterial.js:47]: I think these properties are already created by ShaderMaterial?

Comment (aardgoose) [reply examples/jsm/nodes/materials/NodeMaterial.js:47]: They have the text for a minimal shader applied.
https://github.com/mrdoob/three.js/blob/8d2430fb0a995742844ed667710de43a73d74c54/src/materials/ShaderMaterial.js#L21-L22

setting these to null ensures the new 'shader missing' code in getParameters() is triggered and hence material.onBuild()

**Comment by sunag** [comment]
I think that another fix could be add a `this.type` in `customProgramCacheKey()`.
it's common `WebGLRenderer` users use differents materials just with differents maps and color. Preserve the `cacheKey` can help for it.

image

https://github.com/mrdoob/three.js/blob/8d2430fb0a995742844ed667710de43a73d74c54/examples/jsm/nodes/materials/NodeMaterial.js#L51 

```js
customProgramCacheKey() {

	return this.type + getCacheKey( this );

}
```

**Comment by aardgoose** [comment]
I considered that, downsides it relies on authors coding it consistently and may miss changes of shader text and possible sharing of programs between different materials. Admittedly those may be corner cases, and there may be other consumers of the onBuild() mechanism.

**Comment by sunag** [comment]
Closing in favor of https://github.com/mrdoob/three.js/pull/26227
Thanks!

**Pull request closed without merging**